core/arch/interp.c

```diff
@@ -5771,6 +5771,189 @@ mangle_x64_ib_in_trace(dcontext_t *dcontext, instrlist_t *trace, instr_t *target
+
+static bool
+check_patched_ibl(dcontext_t *dcontext, instrlist_t *trace, instr_t *targeter)
+{
+    int i;
+    instr_t *prev = instr_get_prev(targeter);
+    uint num_instr = prev->length / AARCH64_INSTR_SIZE;
+    /* We start from last instruction to scan for str/ldr pattern. */
+    for (i = num_instr - 2; i >= 0; i--) {
+        uint prev_prev_enc = ((uint *)prev->bytes)[i];
+        uint prev_enc = ((uint *)prev->bytes)[i + 1];
+        /* We bypass the decoder and ompare the raw bytes directly. */
+        if (prev_prev_enc == SAVE_IBL_REG_ENC && prev_enc == RESTORE_IBL_REG_ENC) {
+            return true;
+        }
+    }
+    return false;
+}
+
+/* For AArch64 we have a special case if we cannot remove all code after
+ * the direct branch, which is mangled by cbz/cbnz stolen register.
+ * For example:
+ *    cbz x28, target
+ * would be mangled (see mangle_cbr_stolen_reg() in aarchxx/mangle.c) into:
+ *    str x0, [x28]
+ *    ldr x0, [x28, #32]
+ *    cbnz x0, fall       <- meta instr, not treated as exit cti
+ *    ldr x0, [x28]
+ *    b target            <- delete after
+ * fall:
+ *    ldr x0, [x28]
+ *    b fall_target
+ *    ...
+ * If we delete all code after "b target", then the "fall" path would
+ * be lost. Therefore we need to append the fall path at the end of
+ * the trace as a fake exit stub. Swapping them might be dangerous since
+ * a stub trace may be created on both paths.
+ */
+#    define LDR_TLS_SLOT_MASK 0xffc003e0
+#    define LDR_TLS_SLOT_ENC 0xf9400380
+#    define CBZ_STOLEN_MASK 0x7e000000
+#    define CBZ_STOLEN_ENC 0x34000000
+#    define TBZ_STOLEN_MASK 0x7e000000
+#    define TBZ_STOLEN_ENC 0x36000000
+
+static bool
+instr_is_cbr_stolen(instr_t *instr)
+{
+    if (!instr)
+        return false;
+    else {
+        uint enc = *(uint *)instr->bytes;
+        return ((enc & CBZ_STOLEN_MASK) == CBZ_STOLEN_ENC ||
+                (enc & TBZ_STOLEN_MASK) == TBZ_STOLEN_ENC);
+    }
+}
+
+static bool
+instr_is_load_tls(instr_t *instr)
+{
+    if (!instr || !instr_raw_bits_valid(instr))
+        return false;
+    else {
+        uint enc = *(uint *)instr->bytes;
+        return ((enc & LDR_TLS_SLOT_MASK) == LDR_TLS_SLOT_ENC);
+    }
+}
+
+static instr_t *
+fixup_cbr_on_stolen_reg(dcontext_t *dcontext, instrlist_t *trace, instr_t *targeter)
+{
+    /* We check prior to the targeter whether there is a pattern such as
+     *    cbz/cbnz ...
+     *    ldr reg, [x28, #SLOT]
+     * Otherwise, just return the previous instruction.
+     */
+    instr_t *prev = instr_get_prev(targeter);
+    if (!instr_is_load_tls(prev))
+        return prev;
+    instr_t *prev_prev = instr_get_prev(prev);
+    if (!instr_is_cbr_stolen(prev_prev))
+        return prev;
+    // dr_printf("==============now we confirm ========================\n");
+    // monitor_data_t *md = (monitor_data_t *)dcontext->monitor_field;
+    // instr_disassemble(dcontext, targeter, STDOUT);
```

Comment:
Do we need this commented code? If not, lets remove it.

Reply:
Commented-out code is forbidden in the style guide: https://dynamorio.org/page_code_style.html#autotoc_md125 item 4

Reply:
done